experimental/Origami/src/Origami.jl

```diff
@@ -0,0 +1,333 @@
+include("deck_group.jl")
+include("action.jl")
+include("special_origami.jl")
+include("generate.jl")
+include("homology_action.jl")
+include("cyclic_torus_covers.jl")
+include("homology.jl")
+include("normal_origami.jl")
+include("systoles.jl")
+
+@doc raw"""
+    origami(h::PermGroupElem, v::PermGroupElem)
+
+This function constructs an Origami object from two given permutations h and v . A test is
+performed to check whether the surface described by the given permutations is connected, i.e.
+whether the group generated by the two permutations acts transitively.
+
+# Examples
+```jldoctest
+julia> h = @perm (1,2)
+(1,2)
+
+julia> v = @perm (2,3)
+(1,2)
+
+julia> o = origami(h,v)
+Origami ((1,2),(2,3), 3)
+```
+"""
+function origami(h::PermGroupElem, v::PermGroupElem)
+    d = max(degree(h), degree(v))
+    G = symmetric_group(d)
+    obj = GAP.Globals.Origami(GapObj(h), GapObj(v))
+    return Origami(obj, G(h), G(v), d)
+end
+
+# ugly workaround
+function from_GAP_origami(o::GapObj)
+    d = GAP.Globals.DegreeOrigami(o)::Int
+    G = symmetric_group(d)
+    h = PermGroupElem(G, GAP.Globals.HorizontalPerm(o))
+    v = PermGroupElem(G, GAP.Globals.VerticalPerm(o))
+    return origami(h, v)
+end
+
+function origami_disconnected(h::PermGroupElem, v::PermGroupElem, d::Integer)
+    return Origami(GAP.Globals.OrigamiNC(GapObj(h), GapObj(v), d), h, v, d)
+end
+
+function Base.:(==)(a::Origami, b::Origami)
+    # TODO rewrite this? for now use Gap equality
+    return (a.h == b.h) && (a.v == b.v)
+end
+
+function Base.hash(o::Origami, h::UInt=0x000000000)
+    return hash(o.h, hash(o.v, hash(o.d, h)))
+end
+
+function horizontal_perm(o::Origami)
+    return o.h
+end
+
+function vertical_perm(o::Origami)
+    return o.v
+end
+
+function degree(o::Origami)
+    return o.d
+end
+
+function Base.show(io::IO, o::Origami)
+    h = horizontal_perm(o)
+    v = vertical_perm(o)
+    d = degree(o)
+    print(io, "Origami ($(h),$(v), $(d))")
+end
+
+GapObj(O::Origami) = O.o
+
+@doc raw"""
+    stratum(o::Origami)
+
+This function calculates the stratum of an origami, i.e. the list of the (non-ascendingly
+ordered) non-zero degrees of the singularities. For a singularity of cone angle 2kπ the
+degree of the singularity is k-1.
+
+# Examples
+```jldoctest
+julia> o = origami(cperm([1,2],[3,4]), cperm([2,5,3,6],[4,7]))
+Origami ((1,2)(3,4),(2,5,3,6)(4,7), 7)
+
+julia> stratum(o)
+3-element Vector{Int64}:
+ 2
+ 1
+ 1
+```
+"""
+function stratum(o::Origami)
+    h = horizontal_perm(o)
+    v = vertical_perm(o)
+    commutator = comm(h, v)
+    cycs = cycles(commutator)
+    unsorted_stratum = [length(c) - 1 for c in cycs if length(c) > 1]
+    return sort(unsorted_stratum, rev=true)
+end
+
+@doc raw"""
+    genus(o::Origami)
+
+Calculates the genus of an origami.
+
+# Examples
+```jldoctest
+julia> o = origami(cperm([1,2]), cperm([2,3]))
+Origami ((1,2),(2,3), 3)
+
+julia> genus(o)
+2
+```
+"""
+function genus(o::Origami)
+    return ZZ((sum(stratum(o)) + 2) * 0.5)
```

Comment:
Is the value `sum(stratum(o)) + 2)` going to be always even? I assume it is because otherwise the cast to `ZZ` would error?

In that case it'd be cleaner to write
```suggestion
    return ZZ((sum(stratum(o)) + 2) / 2)
```

Reply:
Yes, it is a property of the stratum that this sum is always even.